### ISSUE TITLE
Reading/Writing to and from buffers

### ISSUE THREAD

**Issue by tmontaigu** [laspy/laspy#78]
Hello,

I'd like to be able to read a Laz / Las file that is stored in a buffer using laspy .
However it doesn't seem possible as the interface only accepts filenames.

This would allow to download a file from an external resource (for example an aws s3 bucket), process the file and then upload it back. 
This could be useful to avoid downloading the file to disk, reading it, rewriting it and then uploading it.

Is it not possible at all currently ?

**Comment by kbevers** [comment]
I don't think so. You are welcome to submit a PR with this feature!

**Comment by tmontaigu** [comment]
I'll try something, but from my understanding of the code, the filename gets propagated from 
laspy.file.File -> laspy.base.FileManager -> laspy.base.DataProvider.
DataProvider being the actual object that opens the file.

So if I modify the DataProvider class to instead use a reference to a file/a buffer , I'll then need to make lots of modifications to 'progate' this change.

Would there be another way to make it work ?

**Comment by davidcaron** [comment]
@tmontaigu 

It was a quick hack I made. I was using FME, a software from Safe. It could provide me with the content of the las file in memory and I wanted to avoid writing it to disk. 

It works, however I can't say I'm happy with the "bytes" parameter and the way it's passed on the way down. It could be cleaner/simpler.

**Comment by tmontaigu** [comment]
@davidcaron 

I had to use the same 'trick' to get something working easily.
I got a branch with this modifications that allows me to read from streams/buffers and get back the
bytes the re-upload/write.

And I agree that it's not clean, however I don't really like the software engeenering of the code.
It makes it harder to have a clean solution. So I ended-up starting something from scratch to see how it could go
